Ks network analysis report 1 (#213)
tari-labs/tari-university: src/network-analysis/probabilistic-attack/MainReport.md

*What is the percentage chance of controlling the majority of nodes in a random sample with varying quantities of the total 
number of nodes, committee size, bad nodes and BFT threshold?*

## Literature Review 

### Tari Digital Assets Network 

The Tari Digital Assets Network (DAN) forms part of the Tari second layer, where the management of all digital 
asset interactions takes place. 

These interactions are processed and validated by committees of special nodes, called Validator Nodes (VNs). Management 
of Digital Assets (DAs) involves state changes and ensures enforcement of the rules that govern assets contracts. 
Thus, all actions on this network are due to the interactions of the VNs. 
The registration of VNs occurs on the base layer. To prevent Sybil attacks, they commit collateral. If 
proved that the VN engaged in malicious behavior, the VN will lose its collateral. 

An Asset Issuer (AI) would then issue DAs and draw up a contract. The AI will dictate the size of 
the committee of VNs for a particular DA. The AI will also have the ability to nominate a trusted node to 
form part of the VN committee for the DA [[14]].

> 

Reviewer: The text could be improved by linking the first use of keywords back to the glossary.

The binomial distribution is the basis for the popular binomial test of statistical significance. 

The binomial distribution is frequently used to model the number of successes in a sample of size $n$ drawn with 
replacement from a population of size $N$. If the sampling is carried out without replacement, the draws are not 
independent, and so the resulting distribution is hypergeometric, not binomial. However, for $n$ much larger than 
$n$, the binomial distribution remains a good approximation and is widely used. 

Thus, in a binomial distribution, an object is selected with replacement [[4]].  A binomial experiment requires that the 
probability of success be constant on every trial. 

**Example:** You have an urn containing $10$ marbles - $5$ red and $5$ green. You randomly select $2$ marbles with replacement, and the 
probability of success would not change. It would be $\frac{5}{10}$ on every trial [[10]]. 

### Monte Carlo Simulations 

The Monte Carlo approach is a computer-based analytical method that was developed in the 1940s as part of the atomic program, by a scientist at the Los Alamos National Laboratory, who used it to 

> model the random diffusion of neutrons.It was named after the city in Monaco and its many casinos. 

Reviewer: ```suggestion
model the random diffusion of neutrons. It was named after the city in Monaco and its many casinos. 
```

are equally likely, each having a probability of $\frac{1}{32}$. Since the first four tosses turn up heads, the probability 
that the next toss is a head is $\frac{1}{2}$ . While a run of five heads has a probability of $\frac{1}{32} = 0.03125$, 
the misunderstanding lies in not realizing that this is the case only before the first coin is tossed. After the first 
four tosses, the results are no longer unknown, so the probability at that point is equal to 1. The reasoning that 
it is more likely that a fifth toss is more likely to be tails because the previous four tosses were heads, with a run 
of luck in the past influencing the odds in the future, forms the basis of the fallacy [[13]].

#### Crude Monte Carlo

The Monte Carlo technique is built upon this principle: instead of evaluating an indefinite integral, which can 
sometimes be impossible, the average of the integrand is estimated, and that is used to approximate the integral. If one 
needs to be more precise, the number of samples can be increased. 

It is a widely used heuristic technique that can solve a variety of common problems including optimization and numerical 

> integration problems. These algorithms work by cleverly sampling from distribution to simulate the workings of a system.  

Reviewer: ```suggestion
integration problems. These algorithms work by cleverly sampling from a distribution to simulate the workings of a system.  
```

## Methodology 
### Notation Used  

This section gives the general notation of statistical expressions when specifically referenced. This information 
is important knowledge for the remainder of the report. 

- Let $N$ be the total number of nodes in the network. 
- Let $n$ be the committee size.
- Let $m$ be the number of bad actors.
- Let $T$ be the BFT threshold.

There is a pool with *N* nodes. The pool contains *m* malicious nodes or bad actors. From within the pool, a random 
selection of nodes, *n* is drawn. From this selection, the probability of drawing a threshold of bad actors, *T*, needs to 
be calculated.  

### Crude Monte Carlo Simulation

#### Programming the Simulation

It was initially thought that selecting a committee ($n$) from the total nodes ($N$) without replacing the selected nodes 

> requires the removal of an element from the pool of total nodes when it is drawn. However, as  the program is 

Reviewer: ```suggestion
requires the removal of an element from the pool of total nodes when it is drawn. However, as the program is 
```

combinations where $B$ is the majority. Thus, from the $10$ combinations, there are $7$ combinations where there is a 
majority of bad nodes. Therefore, the quotient of $7$ and $10$ is the probability $0.7$. 

This method is limited in calculating the probability where the variables are large. For example if the same question was posed, 
but one had to calculate the probability of selecting a majority of bad nodes from a total of $100$ nodes, with a 
committee size of $60$, $60$ bad nodes and $40$ good nodes, the number of combinations where bad and good nodes can be 
chosen is $1.27E+28$.

#### Distribution

##### Binomial Distribution  

With regard to binomial distribution, a committee of nodes is drawn from the total nodes with replacement, i.e. nodes 
are drawn, the intention of the node is distinguished and the node is returned to the total nodes. 
$$
P = {{n}\choose{T}}\cdot\biggl(\frac{m}{n}\biggr)^{T}\cdot\biggl(\frac{N-m}{n}\biggr)^{n-T}
$$

##### Hypergeometric Distribution

With regard to hypergeometric distribution, a committee of nodes is drawn from the total nodes without replacement, 

> i.e. nodes are drawn simultaneously, the intention of the node is distinguished and not returned to the total nodes. This is closely mimics the events that would take place within the network. 

Reviewer: ```suggestion
i.e. nodes are drawn simultaneously, the intention of the node is distinguished and not returned to the total nodes. This closely mimics the events that would take place within the network. 
```

#### Variation of Total Number of Nodes with Committee Size 100

##### Variables

- N = ranging from $0$ to $1300$
- m = $60$% of N
- T = $67$% of N
- n = $100$

<p align="center"><img src="assets/committee_size_100.png" width="700" /></p>
From the two graphs, it can be seen that probabilities are significantly lower when the committee size is 100. There is an increase in probability between $0$ and $700$, albeit, not as steep as the change when the committee size is 10.  The probability plateaus at 0.08.

The larger the committee size, the less dramatic changes there are in the probability. 

#### Variation of Bad Nodes with Committee Size 10 and 100

##### Variables

- N = ranging from $0$ to $50,000$
- m = $10$%, $20$%, $30$%, $40$%, $50$%, $60$%, $70$%, $80$% and $90$% of N
- T = $67$% of N
- n = $10$ and $100$

<p align="center"><img src="assets/bad_actor_grid.png" width="700" /></p>
These graphs show varying probabilities when the number of bad nodes is $20$, $40$, $60$ and $90$. The value when the probability plateaus is used to construct the graph below for both committee sizes $10$ and $100$. 

> <p align="center"><img src="assets/bad_actors_varied_committee_size_10_100.png" width="700" /></p>

Reviewer 2: - This image quality (text) is not good. Do you hacve the original data for the grphs? Possible to make the text 2x-3x larger?
- The x and y axes need descriptions

> The graph shows changes in the probability due to changes in % of bad nodes when the committee size is $10$ and $100$.  When the committee size is 10, there is a change in probability when the bad node percentage is between $30$ and $80$.  When the committee size is 100, there is a steep increase in the probability when the bad node percentage is between $50$ and $80$.  When the committee size is $100$, the probability remains lower as the bad node percentage increases and has a steeper gradient when the change in probability occurs. Whereas, when the committee size is $10$, the probability begins to increase at a lower percentage of bad nodes. 

Reviewer 2: Please ensure all numbers (in report) are consistently encased in $, e.g. one occurence of 10 and 100 each in this paragraph.

Reviewer 2: Please ensure line widths of ~120 are adhered to for easier review comments and comparison of changes - for report in general.
_**Note:** This excludes tables and equations, which should not have newline charcters in between._

## Conclusions and Remarks 

With regards to the Crude Monte Carlo Simulation, at this building block stage, probabilities were calculated and distributions of nodes within the network illustrated.

With regards to the statisical calculation, comments can be made for each of the varied parameters. 

> - Total nodes in the network: the smaller the pool of total nodes in the network, the lower the probability; however, the probability difference is near negligible if the committee size is large. Also, this parameter will be difficult to control, and the network will be ever-increasing 

Reviewer 2: Be specific about which probability is talked about.

> - BFT threshold: this threshold should be higher than $60$%- two thirds ($67$% is ideal)

Reviewer 2: ```suggestion
- BFT threshold: this threshold should be higher than $67$% as per literature.
```
